########## Retrocopy information clean up ##########

# A function to read retrocopy info from Table S1 of Carellie et al. (2016)
read_Carelli2016 <- function(sp = "Human", chroms) {
  require(tidyverse)
  require(readxl)

  # Read excel sheet
  df <- read_excel("data/Carelli2016/Supplemental_Table_S1.xlsx",
    sheet = sp, skip = 5, guess_max = 10000
  )
  # Remove the first NA row in the excel sheet
  df <- df[2:nrow(df), ] %>%
    # change chromsome name style
    dplyr::mutate(
      chr = paste0("chr", chr),
      parental_chr = paste0("chr", parental_chr)
    ) %>%
    # Filter interchromsomal pairs and remove pairs not located on chromosome 1-22 and X
    dplyr::filter(chr != parental_chr, chr %in% chroms, parental_chr %in% chroms) %>%
    dplyr::select(!c(parental_protein, comments))

  names(df) <- c(
    "retroID", "annotatedRetroID", "retroChr", "retroStart", "retroEnd",
    "retroStrand", "parentID", "parentChr", "similarity", "dS"
  )

  return(df)
}


# A function to read retrocopy info from RetrogeneDB2
read_RetrogeneDB2 <- function(retroFile, chroms) {
  require(tidyverse)

  # Read tsv data
  df <- read_tsv(retroFile, guess_max = 10000) %>%
    dplyr::select(
      retrodb_id, external_id, chromosome:strand,
      parental_accession, percent_identity
    ) %>%
    dplyr::mutate(chromosome = paste0("chr", chromosome)) %>%
    dplyr::filter(chromosome %in% chroms)

  names(df) <- c(
    "retroID", "annotatedRetroID", "retroChr", "retroStart", "retroEnd",
    "retroStrand", "parentID", "similarity"
  )

  return(df)
}


# A function to get gene coordinates from gtf file
get_gene_coords <- function(gtf, chroms) {
  require(ensembldb)
  require(dplyr)

  txdb <- makeTxDbFromGFF(gtf)

  gene_df <- as.data.frame(genes(txdb)) %>%
    dplyr::select(!width) %>%
    dplyr::mutate(seqnames = paste0("chr", seqnames)) %>%
    dplyr::filter(seqnames %in% chroms)

  colnames(gene_df) <- c(
    "parentChr", "parentStart", "parentEnd",
    "parentStrand", "parentID"
  )

  return(gene_df)
}


# A function to convert genomic coordinates for chimpanzee and macaque
# liftedR and liftedP are bed files generated by liftover
# retroInfo is a dataframe contains retrocopy info with old coordinates
convert_coords <- function(liftedR, liftedP, retroInfo, chroms) {
  require(tidyverse)

  R <- read.table(text = gsub(":", "\t", gsub("-", "\t", readLines(liftedR)))) %>%
    dplyr::select(V1:V6)

  names(R) <- c(
    "retroChr_lift", "retroStart_lift", "retroEnd_lift",
    "retroChr", "retroStart", "retroEnd"
  )

  P <- read.table(text = gsub(":", "\t", gsub("-", "\t", readLines(liftedP)))) %>%
    dplyr::select(V1:V6)

  names(P) <- c(
    "parentChr_lift", "parentStart_lift", "parentEnd_lift",
    "parentChr", "parentStart", "parentEnd"
  )

  df <- retroInfo %>%
    left_join(R, by = c("retroChr", "retroStart", "retroEnd")) %>%
    inner_join(P, by = c("parentChr", "parentStart", "parentEnd")) %>%
    dplyr::select(
      retroID:annotatedRetroID, retroChr_lift:retroEnd_lift, retroStrand:parentID,
      parentChr_lift:parentEnd_lift, parentStrand, similarity:dS
    ) %>%
    dplyr::filter(
      retroChr_lift != parentChr_lift,
      retroChr_lift %in% chroms, parentChr_lift %in% chroms
    ) %>%
    unique()

  colnames(df) <- gsub("_lift", "", colnames(df), fixed = TRUE)

  return(df)
}


##### Human #####
# Genomic coordinates: hg19

human_chroms <- paste0("chr", c(1:22, "X"))

human_retro <- read_Carelli2016("Human", human_chroms)
human_genes <- get_gene_coords("data/annotations/Homo_sapiens.GRCh37.65.gtf.gz", human_chroms)

human_retro <- human_retro %>% inner_join(human_genes, by = c("parentID", "parentChr"))



##### Mouse #####
# Genomic coordinates: mm9

mouse_chroms <- paste0("chr", c(1:19, "X"))

mouse_retro <- read_Carelli2016("Mouse", mouse_chroms)
mouse_genes <- get_gene_coords("data/annotations/Mus_musculus.NCBIM37.65.gtf.gz", mouse_chroms)

mouse_retro <- mouse_retro %>% inner_join(mouse_genes, by = c("parentID", "parentChr"))



##### Chimpanzee #####
# Genomic coordinates: CHIMP2.1.4

chimp_chroms <- paste0("chr", c(1, 3:22, "X", "2A", "2B"))

chimp_retro <- read_Carelli2016("Chimpanzee", chimp_chroms)
chimp_genes <- get_gene_coords(
  "data/annotations/Pan_troglodytes.CHIMP2.1.4.80.gtf.gz",
  chimp_chroms
)

chimp_retro <- chimp_retro %>% inner_join(chimp_genes, by = c("parentID", "parentChr"))


# Because the genomic coordinates of Chimp HiC data is based on panTro6,
# the coordiantes of Chimp retrocopies and genes need to be converted from CHIMP2.1.4 to panTro6

chimp_retro <- convert_coords(
  "data/Carelli2016/chimp_retro_panTro6.bed",
  "data/Carelli2016/chimp_parent_panTro6.bed",
  chimp_retro, chimp_chroms
)


##### Macaque #####
# Genomic coordinates: MMUL_1

macaque_chroms <- paste0("chr", c(1:19, "X"))

macaque_retro <- read_Carelli2016("Macaque", macaque_chroms)
macaque_genes <- get_gene_coords(
  "data/annotations/Macaca_mulatta.MMUL_1.80.gtf.gz",
  macaque_chroms
)

macaque_retro <- macaque_retro %>% inner_join(macaque_genes, by = c("parentID", "parentChr"))


# Because the genomic coordinates of Macaque HiC data is based on rheMac8,
# the coordiantes of Chimp retrocopies and genes need to be converted from MMUL_1 to rheMac8

macaque_retro <- convert_coords(
  "data/Carelli2016/macaque_retro_rheMac8.bed",
  "data/Carelli2016/macaque_parent_rheMac8.bed",
  macaque_retro, macaque_chroms
)


##### Marmoset #####
# Genomic coordinates: C_jacchus3.2.1

marmoset_chroms <- paste0("chr", c(1:22, "X"))

marmoset_retro <- read_RetrogeneDB2(
  "data/RetrogeneDB2/callithrix_jacchus_retrocopies.tsv",
  marmoset_chroms
)
marmoset_genes <- get_gene_coords(
  "data/annotations/Callithrix_jacchus.C_jacchus3.2.1.80.gtf.gz",
  marmoset_chroms
)

marmoset_retro <- marmoset_retro %>%
  inner_join(marmoset_genes, by = "parentID") %>%
  dplyr::filter(retroChr != parentChr)



##### Cow #####
# Genomic coordinates: UMD3.1

cow_chroms <- paste0("chr", c(1:29, "X"))

cow_retro <- read_RetrogeneDB2("data/RetrogeneDB2/bos_taurus_retrocopies.tsv", cow_chroms)
cow_genes <- get_gene_coords("data/annotations/Bos_taurus.UMD3.1.94.gtf.gz", cow_chroms)

cow_retro <- cow_retro %>%
  inner_join(cow_genes, by = "parentID") %>%
  dplyr::filter(retroChr != parentChr)



##### Dog #####
# Genomic coordinates: CanFam3.1

dog_chroms <- paste0("chr", c(1:38, "X"))

dog_retro <- read_RetrogeneDB2("data/RetrogeneDB2/canis_familiaris_retrocopies.tsv", dog_chroms)
dog_genes <- get_gene_coords(
  "data/annotations/Canis_lupus_familiaris.CanFam3.1.104.gtf.gz",
  dog_chroms
)

dog_retro <- dog_retro %>%
  inner_join(dog_genes, by = "parentID") %>%
  dplyr::filter(retroChr != parentChr)



########## Save data ##########
save.image("Rdata/1-retrocopy_info.Rdata")
